### ISSUE TITLE
Generics for CombinatorialSet (and Operation)

### ISSUE THREAD

**Issue by taogaede** [taogaede/Combinatorics-Package#8]
I definitely intend to look more into generics, but I'm not sure I want to restrict CombinatorialSet instances to containing elements of one particular type.

For example, 
Suppose a CombinatorialSet contains both Integer and Integer[] elements, then one could define something like StrongAddOp, which could add a given Integer both to each Integer and to each entry in the Integer[]s.

Put another way, keeping the type restriction at the element level instead of the set level, allows for more flexibility in how one uses the CombinatorialSet.  Also, math sets don't restrict by type unless defined abstractly with some statement, so it feels a bit more mathy this way.  But again, I haven't thought through everything.  I admit, I can't think of many occasions where one would want to make a CombinatorialSet with a bunch of different types in it, but I don't want to close the door on the option if I don't have to.

I am interested to see if I can simplify the Operation base class using generics though (wildcards look intriguing, though they seem easy to misuse).

**Comment by alexanderbird** [comment]
> then one could define something like StrongAddOp, which could add a given Integer both to each Integer and to each entry in the Integer[]s

Interesting. 

This is something that should be possible to model with a strong type system. The [composite design pattern](https://sourcemaking.com/design_patterns/composite) is a use of polymorphism to treat elements and collections as the same type -- so you can add to a thing without knowing whether that thing is one element or a collection of elements.

One duplicated pattern I'm seeing in this package is the
```
if instanceof String {}
if instanceof Double {}
if instanceof Integer[] {}
....
```
Instead of using these built in types, you could define a Java interface for an `Element` and sets must be composed of Elements. Then you can define concrete `StringElement`, `DoubleElement`, `IntegerElement`, and `CollectionElement` -- where `CollectionElement` is a collection of `Element`s (it could be any of the elements). You would need to decide on what the atomic operations are to support (`add`, `isEqualTo`, etc.?) -- then your `Element` interface would have each of those operations e.g. 

```
interface Element {
    Element add(Element other);
    boolean isEqualTo(Element other);
    boolean isGreaterThan(Element other);
}
```

You'll need to define these operations anyway, but the question is whether you define them inside the concrete element classes or whether you define them throughout the rest of the code.

By the way, I suggest `Element add(Element other)` (immutable operation) rather than `void add(Element other)` (mutating operation that changes the current object) because it makes the code much easier to reason about.

**Comment by taogaede** [comment]
Interfacing.  Cool, I will look into that!  Looks like a neat website.  Thanks.  Would the CollectionElement replace CombinatorialSet in this model?

**Comment by alexanderbird** [comment]
> Cool, I will look into that! Looks like a neat website

Yeah, design patterns were a really important part of my Object Oriented Programming education. That website is helpful, but the best resource on design patterns that I know of is [Design Patterns: Elements of Reusable Object-Oriented Software (1994)](https://www.amazon.ca/Design-Patterns-Elements-Reusable-Object-Oriented/dp/0201633612). I just link to that website because that textbook isn't online for me to link to.

>  Would the CollectionElement replace CombinatorialSet in this model?

Oh interesting, I didn't notice that. I suppose it would -- though I think `CombinatorialSet` is a better name, and you could make `CombinatorialSet` implement `Element`. (So I'm saying CombinatorialSet should replace CollectionElement.)

**Comment by taogaede** [comment]
Okay.  I made a new branch and implemented an Element interface setup, and things appear to be working smoothly.  Operation.java and the Operation subclasses have been simplified a lot.  I think this sort of setup will be a lot more accessible to newcomers.  I'm feeling pretty happy right now lol, thanks so much for suggesting the initial idea!

Using the Element interface idea, I managed to get rid of the input type setting and operate output type checking stuff in the Operation classes - in their place is an Element operation(inputtype input) method.  Anyway, thanks again!

Let me know if you think I've done anything totally crazy and self-sabotaging, otherwise I think I'll merge.

Next step I think will be to organize the printing structure.

**Comment by alexanderbird** [comment]
It looks like a great start! However, as long as you have an if statement that checks a type, you're not done yet ;)

The core idea is that you should be able to do things to an element without caring whether it's a double, and int, a string, or a set. Whatever things can be done to all of those types, those are the things that belong on the Element interface. When you have an if statement that cares about type, you have a concept that belongs inside the element "leaking" outside the element (conceptually) -- we call this a leaky abstraction. Nothing outside of DoubleElement should use the java `Double` type. 

That being said, I would be doing this the same way you are: bit by bit replacing items with their Element counterparts. So I mean it when I say it's a good start -- or at least it's how I start.

As a minor note, you've got elements, sets, and matricies inside each element -- instead I suggest something like:

```java
class DoubleElement implements Element {
     private final Double value;
}
class Set implements Element {
     private final Element[] elements;
}
```

You probably don't need a matrix element because a Set implements Element, and a Set is a set of Elements, so a Set can be a set of Set elements -- a matrix. You won't need a 2D array in any element class.

**Comment by alexanderbird** [comment]
Oh -- regarding the merge -- no reason not to merge IMO, it still works so there's no reason not to if you'd like this in the main branch.
